src/Workspaces/Core/Portable/Indentation/AbstractIndentationService.AbstractIndenter.cs

```diff
@@ -108,8 +115,36 @@ public IndentationResult GetDesiredIndentation(FormattingOptions.IndentStyle ind
                     token, previousNonWhitespaceOrPreprocessorLine, lastNonWhitespacePosition);
             }
 
-            protected abstract IndentationResult GetDesiredIndentationWorker(
-                SyntaxToken token, TextLine previousLine, int lastNonWhitespacePosition);
+            public bool TryGetSmartTokenIndentation(out IndentationResult indentationResult)
+            {
+                indentationResult = default;
+                if (!ShouldUseTokenIndenter(out var token))
+                {
+                    return false;
+                }
+
+                // var root = document.GetSyntaxRootSynchronously(cancellationToken);
+                var sourceText = Tree.GetText(CancellationToken);
+
+                var formatter = CreateSmartTokenFormatter();
+                var changes = formatter.FormatTokenAsync(Document.Project.Solution.Workspace, token, CancellationToken)
+                                       .WaitAndGetResult(CancellationToken);
+
+                var updatedSourceText = sourceText.WithChanges(changes);
+                if (LineToBeIndented.LineNumber < updatedSourceText.Lines.Count)
+                {
+                    var updatedLine = updatedSourceText.Lines[LineToBeIndented.LineNumber];
+                    var offset = updatedLine.GetFirstNonWhitespaceOffset();
+                    if (offset != null)
+                    {
+                        indentationResult = new IndentationResult(
```

Comment:
so .. this can return default? I thought we are changing indentation service to never return default/null.

Reply:
yes.  this def seems like a bug.  nice catch.  i think in this case this would indicate a bug since it would be a case of the indenter saying "i can indent" but then returning null here because it didn't actually update the line in question for that token.  i'll fixup.

Reply:
fixed.